Add Pulsar tutorial (#2892)
quickwit-oss/quickwit: docs/get-started/tutorials/pulsar.md

---
title: Ingest from Pulsar
description: A short tutorial describing how to set up Quickwit to ingest data from Pulsar in a few minutes
tags: [pulsar, integration]
icon_url: /img/tutorials/pulsar.svg
sidebar_position: 5
---

> In this tutorial, we will describe how to set up Quickwit to ingest data from Pulsar in a few minutes. First, we will create an index and configure a Pulsar source. Then, we will create a Pulsar topic and load some events from the [Stackoverflow dataset](link to SO) into it. Finally, we will execute some search.

Reviewer: ```suggestion
In this tutorial, we will describe how to set up Quickwit to ingest data from Pulsar in a few minutes. First, we will create an index and configure a Pulsar source. Then, we will create a Pulsar topic and load some events from the [Stack Overflow dataset](link to SO) into it. Finally, we will execute some searches.
```

## Prerequisites

You will need the following to complete this tutorial:
- A local running [Quickwit instance](/docs/get-started/installation)
- A local running [Pulsar instance](https://pulsar.apache.org/docs/next/getting-started-standalone/). 

### Quickwit setup

[Download](/docs/get-started/installation) Quicwkit and start a server. Then open a new terminal to execute CLI commands with the same binary. 

```bash
./quickwit run
```

Test the cluster is running:

```bash
./quickwit index list
```

### Pulsar setup

```
wget https://archive.apache.org/dist/pulsar/pulsar-2.11.0/apache-pulsar-2.11.0-bin.tar.gz
tar xvfz apache-pulsar-2.11.0-bin.tar.gz
cd apache-pulsar-2.11.0
bin/pulsar standalone

> ```

Reviewer: Maybe worth adding the one command docker run as well here as it's the most convenient method.

Otherwise users may run into the fact that they need java configured and installed among other things to get setup which can be a bit annoying for a tutorial.

## Prepare Quickwit

> First, let's create a new index. Here is the index config and doc mapping corresponding to the schema of stackoverflow posts:

Reviewer: ```suggestion
First, let's create a new index. Here is the index config and doc mapping corresponding to the schema of Stack Overflow posts:
```

```yaml title="index-config.yaml"
#

> # Index config file for stackoverflow dataset.

Reviewer: ```suggestion
# Index config file for Stack Overflow dataset.
```

      type: u64
    - name: creationDate
      type: datetime
      fast: true
      input_formats:
        - rfc3339
      precision: seconds
  timestamp_field: creationDate

search_settings:
  default_search_fields: [title, body]

indexing_settings:
  commit_timeout_secs: 10
```

Execute these Bash commands to download the index config and create the `stackoverflow` index.

```bash
# Download stackoverflow index config.
wget -O stackoverflow.yaml https://raw.githubusercontent.com/quickwit-oss/quickwit/main/config/tutorials/stackoverflow/index-config.yaml

# Create index.
./quickwit index create --index-config stackoverflow.yaml
```

## Create and populate a Pulsar topic

Now, let's create a Pulsar topic and send some messages into it.

```bash
# Execute these commands in the pulsar directory.
bin/pulsar-admin tenants create quickwit
# Check it worked.
bin/pulsar-admin tenants list
bin/pulsar-admin namespaces create quickwit/pulsar
# Set retention policy for the namespace
bin/pulsar-admin namespaces set-retention quickwit/pulsar --time 10m --size 1G
# Create topic.
bin/pulsar-admin topics create-partitioned-topic quickwit/pulsar/stackoverflow -p 4
# Check it worked.

> bin/pulsar-admin topics list-partitioned-topics quickwit/pulsar

Reviewer: I'm tempted to say we should remove this for now for the following reasons:

- We can use the default namespace for this tutorial to get a basic pulsar setup working
- Partitioned topics are not needed to get started although it doesn't do any harm having them here, for simplicity though it's easier to just let pulsar create it for us when we push docs to it.

```python title=send_messages_to_pulsar.py
import json
import pulsar

client = pulsar.Client('pulsar://localhost:6650')
producer = client.create_producer('quickwit/pulsar/stackoverflow')

with open('stackoverflow.posts.transformed-10000.json', encoding='utf8') as file:
   for i, line in enumerate(file):
       producer.send(line.encode('utf-8'))
       if i % 100 == 0:
           print(f"{i}/10000 messages sent.", i)

client.close()
```

Install locally the python client, more details on [documentation page](https://pulsar.apache.org/docs/2.11.x/client-libraries-python/):

```bash
# Requires a python version < 3.11
pip3 install 'pulsar-client==2.10.1'
wget https://raw.githubusercontent.com/quickwit-oss/quickwit/main/config/tutorials/stackoverflow/send_messages_to_pulsar.py
python3 send_messages_to_pulsar.py
```

## Create the Pulsar source

```yaml title="pulsar-source.yaml"
#
# Pulsar source config file.
#
version: 0.4
source_id: pulsar-source
source_type: pulsar
params:
  topics:
    - stackoverflow
  address: pulsar://localhost:6650
```

Run these commands to download the source config file and create the source.

```bash
# Download Pulsar source config.

> wget -O so-pulsar-source.yaml https://raw.githubusercontent.com/quickwit-oss/quickwit/main/config/tutorials/stackoverflow/pulsar-source.yaml

Reviewer: ```suggestion
wget -O stackoverflow-pulsar-source.yaml https://raw.githubusercontent.com/quickwit-oss/quickwit/main/config/tutorials/stackoverflow/pulsar-source.yaml
```

# Create source.

> ./quickwit source create --index stackoverflow --source-config so-pulsar-source.yaml

Reviewer: ```suggestion
./quickwit source create --index stackoverflow --source-config stackoverflow-pulsar-source.yaml
```
For consistency I think it's best to change `so` to `stackoverflow` like everything else

```

> As soon as the Pulsar source is created, Quickwit Control Plane will ask an indexer to start a new indexing pipeline. You can observe that by looking at the logs of indexers. On Kubernetes for example:

Reviewer: Does `Quickwit Control Plane` need to be capitals for `Control` and `Plane` ?

## Time to search!

You can run this command to inspect the properties of the index and check the current number of published splits and documents:

```bash
# Display some general information about the index.
./quickwit index describe --index stackoverflow
```

You are now ready to execute some queries.

```bash
curl 'http://localhost:7280/api/v1/stackoverflow/search?query=search+AND+engine'
```

> It is also possible to access these results through the [Quickwit UI](http://localhost:7280/ui/search?query=&index_id=stackoverflow&max_hits=10).

Reviewer: I like the idea behind this, but if I'm a user having a read through this suggests to me I can click on the link to learn more about the UI rather than expecting it to take me to my localhost (maybe have a separate link for that?)
